src/data/representation_converter.cpp

```diff
@@ -0,0 +1,326 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "data/representation_converter.hpp"
+
+#include "cudf/contiguous_split.hpp"
+#include "data/cpu_data_representation.hpp"
+#include "data/gpu_data_representation.hpp"
+#include "memory/host_table.hpp"
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <algorithm>
+#include <cstring>
+#include <sstream>
+
+namespace cucascade {
+
+// =============================================================================
+// representation_converter_registry implementation
+// =============================================================================
+
+void representation_converter_registry::register_converter_impl(
+  const converter_key& key, representation_converter_fn converter)
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+
+  if (_converters.find(key) != _converters.end()) {
+    std::ostringstream oss;
+    oss << "Converter already registered for source type '" << key.source_type.name()
+        << "' to target type '" << key.target_type.name() << "'";
+    throw std::runtime_error(oss.str());
+  }
+
+  _converters.emplace(key, std::move(converter));
+}
+
+bool representation_converter_registry::has_converter_impl(const converter_key& key) const
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+  return _converters.find(key) != _converters.end();
+}
+
+std::unique_ptr<idata_representation> representation_converter_registry::convert_impl(
+  const converter_key& key,
+  idata_representation& source,
+  const memory::memory_space* target_memory_space,
+  rmm::cuda_stream_view stream) const
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+
+  auto it = _converters.find(key);
+  if (it == _converters.end()) {
+    std::ostringstream oss;
+    oss << "No converter registered for source type '" << key.source_type.name()
+        << "' to target type '" << key.target_type.name() << "'";
+    throw std::runtime_error(oss.str());
+  }
+
+  return it->second(source, target_memory_space, stream);
```

Comment:
It does seem like right now the mutex is hold during the conversion? Maybe we should release the mutex during as we only need it to look up the registry?